libcrux-ml-dsa/src/simd/portable/ntt.rs

```diff
@@ -201,20 +324,81 @@ fn ntt_at_layer_2(re: &mut [Coefficients; SIMD_UNITS_IN_RING_ELEMENT]) {
 }
 
 #[inline(always)]
+#[hax_lib::fstar::before(
+    r#"
+let layer_bound_factor (step_by:usize) : n:nat{n <= 4} =
+    match step_by with
+    | MkInt 1 -> 4
+    | MkInt 2 -> 3
+    | MkInt 4 -> 2
+    | MkInt 8 -> 1
+    | MkInt 16 -> 0
+    | _ -> 0
+"#
+)]
+#[hax_lib::fstar::options("--z3rlimit 600 --split_queries always")]
+#[hax_lib::fstar::before(r#"[@@ "opaque_to_smt"]"#)]
+#[hax_lib::requires(fstar!(r#"
+    (v $STEP_BY > 0) /\
+    (v $OFFSET + v $STEP_BY < v $SIMD_UNITS_IN_RING_ELEMENT) /\
+    (v $OFFSET + 2 * v $STEP_BY <= v $SIMD_UNITS_IN_RING_ELEMENT) /\
+    (Spec.Utils.forall32 (fun i -> (i >= v $OFFSET /\ i < (v $OFFSET + 2 * v $STEP_BY)) ==>
+              Spec.Utils.is_i32b_array_opaque 
+                (v $NTT_BASE_BOUND + ((layer_bound_factor $STEP_BY) * v $FIELD_MAX)) 
+                (Seq.index ${re} i).f_values)) /\
+    Spec.Utils.is_i32b 4190208 $ZETA
+"#))]
+#[hax_lib::ensures(|_| fstar!(r#"
+    Spec.Utils.modifies_range_32 ${re} ${re}_future $OFFSET (${OFFSET + STEP_BY + STEP_BY}) /\
+    (Spec.Utils.forall32 (fun i -> (i >= v $OFFSET /\ i < (v $OFFSET + 2 * v $STEP_BY)) ==>
+              Spec.Utils.is_i32b_array_opaque 
+                (v $NTT_BASE_BOUND + ((layer_bound_factor $STEP_BY + 1) * v $FIELD_MAX)) 
+                (Seq.index ${re}_future i).f_values))
+"#))]
 fn outer_3_plus<const OFFSET: usize, const STEP_BY: usize, const ZETA: i32>(
     re: &mut [Coefficients; SIMD_UNITS_IN_RING_ELEMENT],
 ) {
+    #[cfg(hax)]
+    let orig_re = re.clone();
+
     for j in OFFSET..OFFSET + STEP_BY {
+        hax_lib::loop_invariant!(|j: usize| fstar!(
```

Comment:
It looks like ml-dsa is lacking a `unused_qualifications` warning/deny. Can you add one to `lib.rs`. The qualification here is unused. You should either drop the `use` statement for it, or the `hax_lib` here.

Reply:
fixed, thanks.